### ISSUE TITLE
Import of types nested 2+ levels deep fail to be found when loading schema

### ISSUE THREAD

**Issue by melissag** [dotansimha/graphql-code-generator#4303]
**Describe the bug**
When types are utilized in a "nesting" structure (see sandbox/pasted schema below) 2 or more levels deep, graphql-codegen fails when loading the schema with an `Unknown type` error.  If the nesting is only 1 level deep (i.e.you change the type of `aList` below to be a list of Strings instead of a list of the `nestedB` type), it successfully generates the .ts file.

NOTE that this specifically only happens when the nested types are imported from another file.  If, in the sandbox reproduction, you copy everything out of `user.graphql`, paste it into `schema.graphql`, and remove the import in `schema.graphql`, it works.  However, I'm working on a fairly large/complex project, and so the ability to decompose into multiple component files is key.

If I'm missing something obvious, please let me know!

**To Reproduce**
https://codesandbox.io/s/unknown-type-error-with-2-levels-of-nesting-keqok

1. My GraphQL schema:

*schema.graphql*
```graphql
# import User from 'types/user.graphql'

type Query {
    user(id: ID!): User!
}
```

*user.graphql*
```graphql
type User {
    id: ID!
    username: String!
    email: String!
    nestedFields: [nestedA]
}

type nestedA {
  aList: [nestedB]
}

type nestedB {
  aField: String
}
```

2. My GraphQL operations:
N/A

3. My `codegen.yml` config file:

```yml
schema: schema.graphql
documents: document.graphql
generates:
  types.ts:
    plugins:
      - typescript
      - typescript-operations
```

**Expected behavior**
Same behavior is expected when the types are declared in another file vs. in the main `schema.graphql` file (in other words, the .ts file is generated successfully).

**Environment:**

- OS: Ubuntu 18.04.4
- 
  `"@graphql-codegen/cli": "^1.15.3",`
  `"@graphql-codegen/schema-ast": "^1.15.3",`
  `"@graphql-codegen/typescript": "^1.15.3",`
- NodeJS: 12.x

**Comment by dotansimha** [comment]
Hi @melissag ! Thank you for reporting this. It seems like another issue related to `graphql-import`. 

@ardatan can you please take a look?

**Comment by melissag** [comment]
Any update you can share on this issue?  This is blocking the upgrade of my team's codegen dependencies.

**Comment by ardatan** [comment]
@melissag You can use glob patterns like `**/*.graphql` by disabling import for now if it is urgent.

**Comment by melissag** [comment]
@ardatan I don't think I follow.  For one, I'm not seeing`noGraphQLImport` as an available option: https://graphql-code-generator.com/docs/getting-started/codegen-config; but beyond that, I can't see how disabling import is a tenable option for us.  As called out in the original description, we need the ability to decompose into multiple component files.  But, perhaps I'm missing something in your suggestion?

**Comment by danielrearden** [comment]
Sorry for the confusion @melissag. The option is actually named `skipGraphQLImport`. It's an option that's passed down to the underlying `graphql-tools` library so it's not explicitly called out in the generated documentation. We should definitely make sure we mention it somewhere though. Here's an example usage

```
schema: ./**/*.graphql
generates:
  types.ts:
    plugins:
      - typescript
      - typescript-operations
    config:
      skipGraphQLImport: true
```

With `skipGraphQLImport` set to `false`, codegen supports the `#import` syntax and is supposed to merge all the imported definitions into a single document. This means you should be able to set the `schema` property in your `codegen.yml` to just one "root" file (i.e. schema.graphql) and use `#import` to merge additional files.

However, codegen also supports using a [glob pattern](https://en.wikipedia.org/wiki/Glob_(programming)) for specifying the relevant schema files. So instead of `schema.graphql`, we can specify something like `**/*.graphql`. This makes `#import` mostly obsolete. You can just specify a glob pattern that will match all the relevant `.graphql` files and codegen will combine them just like they'd be combined if you used `#import`.

Note that you can use a glob (and `skipGraphQLImport`) with codegen while still relying on `#import` if you so choose. However, you *could* also use the same glob for loading your schema as well by using `loadSchema` or `loadTypedefs` from `graphql-tools` (which is what codegen uses under the hood).

Ideally, `#import` would just work and you wouldn't need any workarounds. We'll investigate the issue on the `graphql-tools` side of things but hopefully the above workaround can get you unblocked.

**Comment by melissag** [comment]
Thanks so much for this detailed response @danielrearden!  

While I haven't yet been able to get the glob + `skipGraphQLImport` pattern to work, I understand the approach now and believe that it's the complex dependency graph in our project that's causing it to fail, since the order in which the `.graphql` files must be combined is very specific (at least, it is with the way they are defined today).  I'll keep playing with this and see if I can't get it to work, but I have to shift priorities for the immediate future and just wanted to make sure to respond to this.  Thanks again!

**Comment by willpomac** [comment]
I have the same problem and the suggested workaround does not work. Could we get this fixed soon?

**Comment by ardatan** [comment]
@willpomac Did you try to use glob expressions with `skipGraphQLImport`?

**Comment by willpomac** [comment]
![Screen Shot 2021-04-26 at 3 01 47 pm](https://user-images.githubusercontent.com/63992257/116031086-57073380-a6a0-11eb-886d-df8df5338e48.png)

![Screen Shot 2021-04-26 at 3 03 29 pm](https://user-images.githubusercontent.com/63992257/116031235-9766b180-a6a0-11eb-9ad1-2e4b701d1ab0.png)

Yeah I tried --- doesn't work --- see attached images for my codegen.yaml config and also the schema file which has the issue. The `Step` type is not visible at all it seems to the `CreditCard` type

**Comment by willpomac** [comment]
so the problem it appears to be related to @graphql-tools/load and or @graphql-tools/graphql-file-loader.

So spent more time debugging this this time, and even without codegen, graphql loaders break with the nesting....for me anyway, it is not a code gen issue.

Ideally what I want is this:

```
type StepUp {
    type: String!
    mandatory: Boolean!
    url: String!
    sessionId: ID!
}

type PaymentInstruments {
    creditCards: [CreditCard!]!
}

type CreditCard {
    paymentInstrumentId: ID!
    paymentToken: String!
    expiryMonth: String!
    cardSuffix: String!
    scheme: String!
    stepUp: StepUp
}
```

but that doesn't work (StepUp is seen as an unknown type).

so then I did this ugly workaround:

```
type StepUp {
    type: String!
    mandatory: Boolean!
    url: String!
    sessionId: ID!
}

type PaymentInstruments {
    creditCards: [CreditCard!]!

    _stepUp: StepUp  # Not used - to overcome nesting limitation of graphql-import (https://github.com/graphql/graphql-js/issues/2945)
}

type CreditCard {
    paymentInstrumentId: ID!
    paymentToken: String!
    expiryMonth: String!
    cardSuffix: String!
    scheme: String!
    stepUp: StepUp
}
```

**Comment by Urigo** [comment]
thank you for the new info @willpomac 
Do you think you can maybe create a simple reproduction on Code sandbox reproducing this?
Maybe even not on Codegen but on GraphQL Tools?

**Comment by melissag** [comment]
Apologies for my long delay on this.

I concur with @willpomac that it appears to be related to `@graphql-tools/load`.  Though I was able to successfully execute codegen with glob expressions and skipGraphQLImport, when I got to the stage of attempting to load my schema to create my graphQL handler, I encountered the same `Unknown type` failure.

@ardatan It seems highly likely that this is related to (or potentially a duplicate of) [graphql-tools/issues/1591](https://github.com/ardatan/graphql-tools/issues/1591), and I notice that you just released a fix for that.  Can you keep us posted on when the `graphql-tools` dependency is updated in this repo?  Keeping fingers crossed that it resolves this issue as well.

**Comment by melissag** [comment]
Yup, confirmed that the latest version of `@graphql-codegen/cli` (1.21.4) completely fixes this issue.
